LLL: end to end testing

zigguratt opened the pull request:
Added tests for `panic` and `exp`.

zigguratt committed:
Added 'panic' tests.

zigguratt committed:
Added tests for exponent operator.

zigguratt committed:
Simplified exp_operator_on_range test and changed to tedting a range.

zigguratt committed:
Removed enclosed_panic test until assembler bug is fixed.

zigguratt committed:
Added tests for LLL constructors. Fixed a few tab issues.

axic commented:
Please use `git rebase`, there should not be any merge commits in pull requests.

zigguratt commented:
Merge commit removed

axic commented on test/liblll/EndToEndTest.cpp:
```diff
@@ -38,9 +38,93 @@ BOOST_FIXTURE_TEST_SUITE(LLLEndToEndTest, LLLExecutionFramework)
-        char const* sourceCode = "(returnlll { (return \"test\") })";
-        compileAndRun(sourceCode);
-        BOOST_CHECK(callFallback() == encodeArgs(string("test", 4)));
+	char const* sourceCode = "(returnlll { (return \"test\") })";
```
I think you're changing indentation throughout the file?

axic commented on test/liblll/EndToEndTest.cpp:
```diff
@@ -43,6 +43,90 @@ BOOST_AUTO_TEST_CASE(smoke_test)
+						(return @@0x00))
+					(when (= (div (calldataload 0x00) (exp 2 224)) 0x89ea642f)
+						(return @@0x01))
+					(jump 0x02))))
```
If there's a "panic" here an appropriate test should be added `callFallback()` and even `callContractFunction("invalidFunction()")`.

Probably though this should split into two tests:
- constructor argument saving (this current one, which doesn't need a `panic`)
- `when` waterfall

axic commented:
We should also have tests for the arithmetic/binary/unary operations:
```
                std::map<std::string, Instruction> const c_arith = {
                        { "+", Instruction::ADD },
                        { "-", Instruction::SUB },
                        { "*", Instruction::MUL },
                        { "/", Instruction::DIV },
                        { "%", Instruction::MOD },
                        { "&", Instruction::AND },
                        { "|", Instruction::OR },
                        { "^", Instruction::XOR }
                };
                std::map<std::string, pair<Instruction, bool>> const c_binary = {
                        { "<", { Instruction::LT, false } },
                        { "<=", { Instruction::GT, true } },
                        { ">", { Instruction::GT, false } },
                        { ">=", { Instruction::LT, true } },
                        { "S<", { Instruction::SLT, false } },
                        { "S<=", { Instruction::SGT, true } },
                        { "S>", { Instruction::SGT, false } },
                        { "S>=", { Instruction::SLT, true } },
                        { "=", { Instruction::EQ, false } },
                        { "!=", { Instruction::EQ, true } }
                };
                std::map<std::string, Instruction> const c_unary = {
                        { "!", Instruction::ISZERO },
                        { "~", Instruction::NOT }
                };
```

pirapira commented:
Restarting appveyor.

zigguratt committed:
Removed 'panic' from 'constructor_arguments_external' and added specific
test for fallback and fallthrough. Added one test for 'lit'. Added tests
for arithmetic/binary/unary operations. Added the start of tests for
assembler instructions within LLL source.

axic commented on test/liblll/EndToEndTest.cpp:
```diff
@@ -43,6 +43,235 @@ BOOST_AUTO_TEST_CASE(smoke_test)
+		(returnlll
+			(seq
+				(mstore8 0x00 (! (< 53 87)))
+				(mstore9 0x01 (! (>= 42 73)))
```
`mstore9` :)
